torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +858,334 @@ def _get_ignored_params(
+        device_from_device_id: Optional[torch.device],
+    ) -> None:
+        """
+        Materializes the wrapped module ``module`` in place if needed: either
+        if the module has parameters that use meta device or are torchdistX
+        fake tensors.
+
+        This method uses ``param_init_fn`` to materialize the module if the
+        function is not ``None`` and falls back to default behavior otherwise.
+        For meta device, this moves the module to ``device_from_device_id`` if
+        it is not ``None`` or the current device otherwise and calls
+        ``reset_parameters()``, and for torchdistX fake tensors, this calls
+        ``deferred_init.materialize_module()``.
+        """
+        is_meta_module = any(p.is_meta for p in module.parameters())
+        is_torchdistX_deferred_init = (
+            not is_meta_module
+            and _TORCHDISTX_AVAIL
+            and any(fake.is_fake(p) for p in module.parameters())
+        )
+        if (
+            is_meta_module or is_torchdistX_deferred_init
+        ) and param_init_fn is not None:
+            if not callable(param_init_fn):
+                raise ValueError(
+                    f"Expected {param_init_fn} to be callable but got {type(param_init_fn)}"
+                )
+            param_init_fn(module)
+        elif is_meta_module:
+            # Run default meta device initialization
+            materialization_device = device_from_device_id or torch.cuda.current_device()
+            module.to_empty(device=materialization_device)
+            try:
+                with torch.no_grad():
+                    module.reset_parameters()
+            except BaseException as e:
+                warnings.warn(
+                    "Unable to call `reset_parameters()` for module on meta "
+                    f"device with error {str(e)}. Please ensure your "
+                    "module implements a `reset_parameters()` method."
+                )
+                raise e
+        elif is_torchdistX_deferred_init:
+            # Run default torchdistX initialization
+            deferred_init.materialize_module(
+                module,
+                check_fn=lambda k: not isinstance(k, FullyShardedDataParallel),
+            )
+
+    def _move_module_to_device(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+        device_from_device_id: Optional[torch.device],
+    ):
+        """
+        Moves ``module`` depending on ``device_from_device_id`` and its current
+        device.
+
+        - If ``device_from_device_id`` is not ``None``, then this moves
+        ``module`` to the device.
+        - If ``device_from_device_id`` is ``None``, then this does not move
+        ``module`` but warns the user if it is on CPU.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        cpu_device = torch.device("cpu")
+        param = next(self._get_orig_params(module, ignored_params), None)
```

Comment:
Any way this could throw with a StopIteration instead of returning None?

Reply:
`_get_orig_params()` has a `try: ... except StopIteration: pass`, so it should not throw `StopIteration`. My understanding is that once the `Iterator` returns `None`, it ends, so that the `StopIteration` -> `pass` will lead to returning `None` and breaking out of the `Iterator`.

Reply:
```
import torch
seq = torch.nn.Sequential()  # no parameters
def _get_orig_params(
    module,
):   
    param_gen = module.parameters()
    try:
        while True:
            param = next(param_gen)
            yield param
    except StopIteration:
        pass
next(_get_orig_params(seq), None)  # returns `None`
```

From the documentation for `next()`:
> Return the next item from the iterator. If default is given and the iterator is exhausted, it is returned instead of raising StopIteration.
By specifying `None` as the default for `next()`, it should never raise `StopIteration`.